### PR TITLE
Cog 337 llama index support

### PR DESCRIPTION
**Pull request by dexters1**
## Release Notes

- **New Features**
	- Introduced the `ingest_data_with_metadata` function for ingesting data along with metadata.
	- Added the `save_data_item_to_storage` function to handle various data item types for storage.
	- Created a new Jupyter notebook for utilizing LlamaIndex documents within the Cognee framework.
	- Added the `save_data_item_with_metadata_to_storage` function for enhanced data storage capabilities.

- **Improvements**
	- Updated error messages in permission checks to use user email instead of username.
	- Streamlined the `save_data_to_storage` function for improved efficiency.
	- Simplified parameter passing in the `cognee.search` function.

- **Dependencies**
	- Updated version constraints for `instructor` and `tenacity` packages.
	- Added new dependency `llama-index-core`.
	- Updated `psycopg2` to include an optional attribute.

- **Chores**
	- Added a new GitHub Actions workflow for testing the LlamaIndex notebook.

### CODE REVIEW

File: cognee/modules/users/permissions/methods/check_permission_on_documents.py [hunk 1]
```diff
@@ -33,4 +33,4 @@ async def check_permission_on_documents(user: User, permission_type: str, docume
         has_permissions = all(document_id in resource_ids for document_id in document_ids)
 
         if not has_permissions:
-            raise PermissionDeniedException(f"User {user.username} does not have {permission_type} permission on documents")
+            raise PermissionDeniedException(f"User {user.email} does not have {permission_type} permission on documents")
```
Comment (0xideas) [inline cognee/modules/users/permissions/methods/check_permission_on_documents.py:36]: what's this change about?

File: cognee/tasks/ingestion/ingest_data_with_metadata.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            # Ingest data and add metadata
+            with open(file_path.replace("file://", ""), mode = "rb") as file:
+                classified_data = ingestion.classify(file)
+
+                data_id = ingestion.identify(classified_data)
+
+                file_metadata = classified_data.get_metadata()
+
+                from sqlalchemy import select
+                from cognee.modules.data.models import Data
```
Comment (0xideas) [inline cognee/tasks/ingestion/ingest_data_with_metadata.py:40]: What's the reason for these imports being within the function?

File: notebooks/cognee_llama_index.ipynb [hunk 3]
```diff
@@ -0,0 +1,205 @@
+{
```
Comment (0xideas) [inline notebooks/cognee_llama_index.ipynb:1]: Should we add a test analogous to [test_notebook](https://github.com/topoteretes/cognee/blob/main/.github/workflows/test_notebook.yml) for this new notebook?

**Review by 0xideas — COMMENTED** [review]
Looks really good overall, I left a few questions/suggestions. Can you think of any documentation that we should update/expand?

File: cognee/tasks/ingestion/ingest_data_with_metadata.py (hunk 2 above)
Comment (dexters1) [reply cognee/tasks/ingestion/ingest_data_with_metadata.py:40]: I've inherited that part of the code from ingest_data.py so I'm not sure

File: notebooks/cognee_llama_index.ipynb (hunk 3 above)
Comment (0xideas) [reply notebooks/cognee_llama_index.ipynb:1]: @dexters1 , what do you think?

Comment (dexters1) [reply notebooks/cognee_llama_index.ipynb:1]: We can, if we plan on having more notebooks in the future we can make my test_notebook action reusable as there will be little difference between notebooks

Comment (dexters1) [reply notebooks/cognee_llama_index.ipynb:1]: I've added a test for this notebook in this MR, but the reusable action will be made in the next sprint. I don't have time to take care of it in this sprint.

File: notebooks/cognee_llama_index.ipynb [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from cognee import SearchType\n",
+    "\n",
+    "# Query cognee for summaries\n",
+    "search_results = await cognee.search(\n",
+    "    SearchType.SUMMARIES,\n",
+    "    {'query': \"What are the main news discussed in the document?\"}\n",
```
Comment (borisarzentar) [inline notebooks/cognee_llama_index.ipynb:153]: `query_text` should be a text, not dict with `query` prop.

File: .github/workflows/test_cognee_llama_index_notebook.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
+name: test | llama index notebook
+
+on:
+  workflow_dispatch:
+  pull_request:
+    branches:
+      - main
+    types: [labeled]
```
Comment (0xideas) [inline .github/workflows/test_cognee_llama_index_notebook.yml:8]: currently, we apply [labeled, synchronize] in the other GitHub actions

**Review by 0xideas — COMMENTED** [review]
Looks good to me, except for the small thing in GitHub actions, for consistency's sake :)

**Comment by 0xideas** [comment]
(@dexters1 could you also merge the latest main?) <- actually, never mind, its not necessary!

File: pyproject.toml [hunk 6]
```diff
@@ -69,7 +69,8 @@ fastapi-users = {version = "*", extras = ["sqlalchemy"]}
 alembic = "^1.13.3"
 asyncpg = "^0.29.0"
 pgvector = "^0.3.5"
-psycopg2 = "^2.9.10"
+psycopg2 = {version = "^2.9.10", optional = true}
+llama-index-core = "^0.11.22"
```
Comment (Vasilije1990) [inline pyproject.toml:73]: Make this optional please

File: pyproject.toml [hunk 7]
```diff
@@ -36,7 +36,7 @@ boto3 = "^1.26.125"
 botocore="^1.35.54"
 gunicorn = "^20.1.0"
 sqlalchemy = "2.0.35"
-instructor = "1.6.3"
+instructor = "^1.5.2"
```
Comment (Vasilije1990) [inline pyproject.toml:39]: why not fix the version? it is not a very reliable lib, as ti changed a few times already

File: .github/workflows/test_cognee_llama_index_notebook.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+jobs:
+  get_docs_changes:
+    name: docs changes
+    uses: ./.github/workflows/get_docs_changes.yml
+
+  run_notebook_test:
+    name: test
+    needs: get_docs_changes
+    if: needs.get_docs_changes.outputs.changes_outside_docs == 'true' && ${{ github.event.label.name == 'run-checks' }}
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (dexters1) [reply .github/workflows/test_cognee_llama_index_notebook.yml]: @borisarzentar  Added change, maybe we should change this for the rest of the actions as well?

**Review by borisarzentar — APPROVED** [review]

**Pull request merged**